src/FESpaces/ConformingFESpaces.jl

```diff
@@ -139,6 +141,119 @@ function CellFE(
     max_order)
 end
 
+function get_dof_basis(model::DiscreteModel,
+                       cell_reffe::AbstractArray{<:ReferenceFE})
+  lazy_map(get_dof_basis,
+           cell_reffe,
+           get_cell_map(Triangulation(model)))
+end
+
+function get_dof_basis(model::DiscreteModel,
```

Comment:
This includes a dependency between discrete model and RaviartThomas reffes.  As a result, RaviartThomas reffes cannot be fully implemented in the ReferenceFEs module, which is sort of confusing. It is also a sign that the current API of ReferenceFEs is not general enough.

Do you think it's possible to fix this? Perhaps it is possible with some refactoring of the ReferenceFEs API.

Reply:
See https://github.com/gridap/Gridap.jl/pull/579#discussion_r621828947

Reply:
I have been talking to @amartinhuertas this morning, and I think this is the way to go. Div-conforming FE spaces require to modify the reference basis when transformed to the local FE basis. That is the identity for Lagrangian FEM (grad-conforming) but not the case (in general) for other FEs like div or curl-conforming ones.

This is something that cannot be at the ref FE level (it is a global thing, to enforce div-conformity in the global space).

I also don't like to have dispatching wrt RT in the `ConformingFESpace` file. We have also learned that the __only__ modification that is required to go to div-conforming are these two functions that compute the global DOF and shape functions on physical cells. So, it seems to me that the most neat solution is to create a file for these FE spaces that simply implement these functions.

It is in the same spirit as e.g. discontinuous Galerkin and conforming FE spaces, which fill call the unconstrained FE constructor in different ways. We could also do the dispatching at that level for div-conforming, but since the code is the same with only the previous exception, I would just implement the functions that need dispatching.

Cheers,

--Santi

Reply:
We can deal with this If it is necessary and I would also reame the versions taking the discrete model to `get_cell_shapefuns` and `get_cell_dof_basis` like other functions returning cell-wise values.

Back in the day, `get_shapefuns` and `get_dof_basis` had only a single argument (the reference FE), but we included a second argument (the geometrical map) to handle RT and Nedelec. Now these FE types will be handled differently, so perhaps we can revert to the original definition of `get_shapefuns` and `get_dof_basis` with just a single argument.  This would simplify the implementation of Lagrangian reffes, which became a bit dirty after introducing RT and Nedelec.

In summary, these would the API

```julia
# Abstract methods in ReferenceFEs module
get_shapefuns(reffe::ReferenceFE)
get_dof_basis(reffe::ReferenceFE)

# Methods with a default implementation in terms of previous ones (which will work for e.g. Lagrangian),
# but they can be overloaded by new ReferenceFE types
get_cell_shapefuns(model::DiscreteModel,cell_reffe::AbstractArray{<:ReferenceFE}))
get_cell_dof_basis(model::DiscreteModel,cell_reffe::AbstractArray{<:ReferenceFE}))

# Methods to be added for RT FEs

# 1) Helper methods specific for RT
get_shapefuns(reffe::GenericRefFE{RaviartThomas},map::Field,signflip::Integer)
get_dof_basis(reffe::GenericRefFE{RaviartThomas},map::Field,signflip::Integer)

# 2) Specialization of Abstract methods in ReferenceFEs module
get_shapefuns(reffe::GenericRefFE{RaviartThomas}) = get_shapefuns(reffe,GenericField(identity),1)
get_dof_basis(reffe::GenericRefFE{RaviartThomas}) = get_dof_basis(reffe,GenericField(identity),1)

# 3) Specialization of cell-wise (like already implemented in the PR)
get_cell_shapefuns(model::DiscreteModel,cell_reffe::AbstractArray{<:GenericRefFE{RaviartThomas}}))
get_cell_dof_basis(model::DiscreteModel,cell_reffe::AbstractArray{<:GenericRefFE{RaviartThomas}}))
```

What do you think?

Reply:
Agreed

Reply:
Btw, the sign is not +1 or -1, not an `Int` but an array with as many +1/-1 as DOFs in the RefFE

Reply:
Ok! I have updated the code snipped

Reply:
I used `Bool`, not `Int`. (true-> -1 false -> 1; not a big deal)

Reply:
Done in 1685e4d99808277dba2e2f7062deb14e6f969bdb